### ISSUE TITLE
dev-build command : get the spack-xxx directories created out of source 

### ISSUE THREAD

**Issue by aphecetche** [LLNL/spack#21571]
Currently, when doing  :
``` 
cd some_place
spack dev-build -d location_of_source --drop-in zsh some_package  
```
spack creates a few `spack-xxx` files and a directory (`spack-build-yyy`) that are located under `location_of_source`, i.e. "in source". 

```
$ cd some_package && ls -1dF spack-*
spack-build-env.txt
spack-build-out.txt
spack-build-zp32c65/
spack-configure-args.txt
```

Note that in my use case `some_package` is a CMakePackage, hence capable of out-of-source builds (from the `spack-build-zp32c65/` in this example). So technically speaking the build is out-of-source, but the directory is in-source (hope my description makes sense).

### Rationale

Keep the source directory clean.

### Description

I would like to have the option to get those files/directories created in `some_place` (i.e. the directory from where the spack dev-build command was launched) instead of "polluting" the source directory. And/or have an option in the `dev-build` command to choose the destination directory.

### General information

* **Spack:** 0.16.0-951-3893ce8f71
* **Python:** 3.8.2
* **Platform:** darwin-bigsur-skylake
* **Concretizer:** original

**Comment by haampie** [comment]
Yeah, you're not the only one who wants to have more control over the build directory. See also https://github.com/spack/spack/pull/18838. Pinging @becker33 for some input.

You can however override the build directory at the package.py level, which gives you some control. At the same time if you fix that to say `build/`, `spack dev-build` will use the same build folder for different specs, which is also something you might not want.

Would be useful if we had these flags `spack dev-build --build <dir a> --source <dir b> <spec>`; kinda like `cmake --build <dir a> <dir b>`. Same for environments:

```yaml
spack:
  specs: [mypkg@x.y.z]
  develop:
    - spec: mypkg@x.y.z
      source: path/to/sources
      build: path/to/bindir
```

See also https://github.com/spack/spack/issues/19359.

What I don't know is what should happen in case the build system (generator) does not support out of source builds.

**Comment by becker33** [comment]
I'm working on this right now. What will happen if the build system does not support out of source builds? The build will fail. For now, we will just put it on the user to only use this feature where it will work. Some autotools packages support out-of-source builds. Some cmake packages don't allow the out-of-source build directory to be modified (not sure how they even manage that in CMake, but...).

**Comment by psakievich** [comment]
Should we close this? The feature has been implemented.
